src/functional/sim_piece.zig

```diff
@@ -0,0 +1,765 @@
+                    current_metadata.upper_bound_slope,
+                );
+            } else {
+                // The current segment metadata cannot be merged. Append the merged metadata and
+                // create a new one.
+                for (timestamps_array.items) |timestamp| {
+                    try merged_segments_metadata.append(.{
+                        .start_time = timestamp,
+                        .interception = merge_metadata.interception,
+                        .lower_bound_slope = merge_metadata.lower_bound_slope,
+                        .upper_bound_slope = merge_metadata.upper_bound_slope,
+                    });
+                }
+                timestamps_array.clearRetainingCapacity();
+                merge_metadata = .{
+                    .start_time = 0.0,
+                    .interception = current_metadata.interception,
+                    .lower_bound_slope = current_metadata.lower_bound_slope,
+                    .upper_bound_slope = current_metadata.upper_bound_slope,
+                };
+                try timestamps_array.append(current_metadata.start_time);
+            }
+        }
+
+        if (timestamps_array.items.len > 0) {
+            // Append the final merged segment metadata.
+            for (timestamps_array.items) |timestamp| {
+                try merged_segments_metadata.append(.{
+                    .start_time = timestamp,
+                    .interception = merge_metadata.interception,
+                    .lower_bound_slope = merge_metadata.lower_bound_slope,
+                    .upper_bound_slope = merge_metadata.upper_bound_slope,
+                });
+            }
+            timestamps_array.clearRetainingCapacity();
+        }
+    }
+
+    // This step is needed since the timestamp order is lost due to the hashmap.
+    mem.sort(
+        SegmentMetadata,
+        merged_segments_metadata.items,
+        {},
+        compareMetadataByStartTime(),
+    );
+}
+
+/// Sim-Piece Phase 3. Compute a hash map from interception points in `merged_segments_metadata`
+/// to a hash map from the approximation slope to an array list of timestamps and store in
+/// `merged_segments_metadata_map`. The `allocator` is used to allocate memory of intermediates.
+fn computeSegmentsMetadataHashMap(
+    merged_segments_metadata: ArrayList(SegmentMetadata),
+    merged_segments_metadata_map: *HashMap(f64, HashMap(
+        f64,
+        ArrayList(usize),
+        HashF64Context,
+        std.hash_map.default_max_load_percentage,
+    ), HashF64Context, std.hash_map.default_max_load_percentage),
+    allocator: mem.Allocator,
+) !void {
+    for (merged_segments_metadata.items) |segment_metadata| {
+        const interception: f64 = segment_metadata.interception;
+        const slope: f64 = (segment_metadata.lower_bound_slope +
+            segment_metadata.upper_bound_slope) / 2;
+
+        // Get or put the outer HashMap entry for the given interception
```

Comment:
Missing . at the end. Although maybe either delete the comment or extend it to state why it is done as it is clear a get put method is executed from the code.

Reply:
Fixed.